clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -850,54 +952,51 @@ synchronized boolean hasOngoingTransaction() {
+        // numbers of in-flight batches from sequence 0, then retry the failed batch, which should now succeed.
+        // For the transactional producer, allow the batch to fail. When processing the failed batch, we will
+        // transition to an abortable error and set a flag indicating that we need to bump the epoch
+        if (error == Errors.UNKNOWN_PRODUCER_ID && !isTransactional()) {
+            requestEpochBumpForPartition(batch.topicPartition);
+            return true;
+        } else if (error == Errors.OUT_OF_ORDER_SEQUENCE_NUMBER) {
+            if (!hasUnresolvedSequence(batch.topicPartition) &&
+                    (batch.sequenceHasBeenReset() || !isNextSequence(batch.topicPartition, batch.baseSequence()))) {
+                // We should retry the OutOfOrderSequenceException if the batch is _not_ the next batch, ie. its base
+                // sequence isn't the lastAckedSequence + 1.
                 return true;
-            } else if (lastAckedOffset(batch.topicPartition).orElse(NO_LAST_ACKED_SEQUENCE_NUMBER) < response.logStartOffset) {
-                // The head of the log has been removed, probably due to the retention time elapsing. In this case,
-                // we expect to lose the producer state. Reset the sequences of all inflight batches to be from the beginning
-                // and retry them.
-                startSequencesAtBeginning(batch.topicPartition);
+            } else if (!isTransactional()) {
+                // For the idempotent producer, retry all OUT_OF_ORDER_SEQUENCE_NUMBER errors. If there are no
+                // unresolved sequences, or this batch is the one immediately following an unresolved sequence, we know
+                // there is actually a gap in the sequences, and we bump the epoch. Otherwise, retry without bumping
+                // and wait to see if the sequence resolves
+                if (!hasUnresolvedSequence(batch.topicPartition) ||
+                        isNextSequenceForUnresolvedPartition(batch.topicPartition, batch.baseSequence())) {
+                    requestEpochBumpForPartition(batch.topicPartition);
+                }
                 return true;
             }
         }
-        return false;
+
+        // If neither of the above cases are true, retry if the exception is retriable
+        return error.exception() instanceof RetriableException;
     }
 
     // visible for testing
     synchronized boolean isReady() {
         return isTransactional() && currentState == State.READY;
     }
 
+    void handleCoordinatorReady() {
+        NodeApiVersions nodeApiVersions = transactionCoordinator != null ?
+                apiVersions.get(transactionCoordinator.idString()) :
+                null;
+        this.coordinatorSupportsBumpingEpoch = nodeApiVersions != null &&
+                nodeApiVersions.apiVersion(ApiKeys.INIT_PRODUCER_ID).maxVersion >= 3;
```

Comment:
`nodeApiVersions.apiVersion(ApiKeys.INIT_PRODUCER_ID)` could be null?

Reply:
Good call, thanks. Really wishing for a null-coalescing operator here.